jlm/llvm/opt/ScalarEvolution.cpp

```diff
@@ -422,35 +289,247 @@ ScalarEvolution::CreateChainRecurrences(
+      }
+
+      // Find the start value for the recurrence
+      if (auto simpleNode =
+              rvsdg::TryGetOwnerNode<const rvsdg::SimpleNode>(*loopVar.input->origin()))
+      {
+        if (rvsdg::is<IntegerConstantOperation>(simpleNode->GetOperation()))
+        {
+          // If the input value is a constant, get it's SCEV representation
+          chainRecurrence->SetStartValue(
+              GetOrCreateSCEVForOutput(*loopVar.input->origin())->Clone());
+        }
+      }
+      else
+      {
+        // If not, create a SCEVInit node representing the start value
+        chainRecurrence->SetStartValue(std::make_unique<SCEVInit>(*loopVarPre));
+      }
+      ChainRecurrenceMap_.emplace(loopVarPre, std::move(chainRecurrence));
+    }
+    else
+    {
+      auto unknownChainRecurrence = std::make_unique<SCEVChainRecurrence>(thetaNode);
+      unknownChainRecurrence->AddOperand(std::make_unique<SCEVUnknown>());
+      ChainRecurrenceMap_.emplace(loopVarPre, std::move(unknownChainRecurrence));
+    }
   }
 
-  std::unordered_map<const rvsdg::Output *, std::unique_ptr<SCEV>> chrecMap{};
-  for (auto indVar : inductionVariableCandidates.Items())
-    chrecMap[indVar] = UniqueSCEVs_.at(indVar)->Clone();
+  std::unordered_map<const rvsdg::Output *, std::unique_ptr<SCEVChainRecurrence>> chrecMap{};
+  for (const auto loopVar : thetaNode.GetLoopVars())
+  {
+    auto storedRec = ChainRecurrenceMap_.at(loopVar.pre)->Clone();
+    // Workaround for the fact that Clone() is an overrided method that returns a unique_ptr of SCEV
+    chrecMap[loopVar.pre] = std::unique_ptr<SCEVChainRecurrence>(
+        dynamic_cast<SCEVChainRecurrence *>(storedRec.release()));
+  }
 
   return chrecMap;
 }
 
+std::unique_ptr<SCEVChainRecurrence>
+ScalarEvolution::CreateChainRecurrence(
+    const rvsdg::Output & IV,
+    const SCEV & scevTree,
+    const rvsdg::ThetaNode & thetaNode)
+{
+  auto chrec = std::make_unique<SCEVChainRecurrence>(thetaNode);
+
+  if (const auto scevConstant = dynamic_cast<const SCEVConstant *>(&scevTree))
+  {
+    // This is a constant, we add it as the only operand
+    chrec->AddOperand(scevConstant->Clone());
+  }
+  else if (const auto scevPlaceholder = dynamic_cast<const SCEVPlaceholder *>(&scevTree))
+  {
+    if (scevPlaceholder->GetPrePointer() == &IV)
+    {
+      // Since we are only interested in the step value, and not the initial value, we can ignore
+      // ourselves by adding a 0, which is the identity element
+      chrec->AddOperand(std::make_unique<SCEVConstant>(0));
+    }
+    else
+    {
+      if (ChainRecurrenceMap_.find(scevPlaceholder->GetPrePointer()) != ChainRecurrenceMap_.end())
+      {
+        // We have a dependency of another IV
+        // Get it's saved value. This is safe to do due to the topological ordering
+        auto storedRec = ChainRecurrenceMap_.at(scevPlaceholder->GetPrePointer())->Clone();
+
+        return std::unique_ptr<SCEVChainRecurrence>(
```

Comment:
This return statement here is a bit surprising, as all other branches use the `chrec` to create the return value.